lib/view_component/slotable_v2.rb

```diff
@@ -0,0 +1,222 @@
+# frozen_string_literal: true
+
+require "active_support/concern"
+
+require "view_component/slot"
+
+module ViewComponent
+  module Slotable
+    ##
+    # Version 2 of the Slots API
+    module V2
+      extend ActiveSupport::Concern
+
+      # Setup component slot state
+      included do
+        # Hash of registered Slots
+        class_attribute :registered_slots
+        self.registered_slots = {}
+      end
+
+      class_methods do
+        ##
+        # Registers a slot on the component.
+        #
+        # = Example
+        #
+        #   renders_one :header do
+        #     def initialize(classes:)
+        #       @name = name
+        #     end
+        #   end
+        #
+        # = Rendering slot content
+        #
+        # The component's sidecar template can access the slot by calling a
+        # helper method with the same name as the slot (pluralized if the slot
+        # is a collection).
+        #
+        #   <h1>
+        #     <%= header %>
+        #   </h1>
+        #
+        # = Setting slot content
+        #
+        # Renderers of the component can set the content of a slot by calling a
+        # helper method with the same name as the slot. For collection
+        # components, the method can be called multiple times to append to the
+        # slot.
+        #
+        #   <%= render_inline(MyComponent.new) do |component| %>
+        #     <%= component.header(classes: "Foo") do %>
+        #       <p>Bar</p>
+        #     <% end %>
+        #   <% end %>
+        def renders_one(slot_name, callable = nil)
```

Comment:
Might we name `callable` to be more domain-specific to rendering? How about `renderable`?

```suggestion
        def renders_one(slot_name, renderable = nil)
```